### ISSUE TITLE
build fails if backend compiler has ssp headers installed

### ISSUE THREAD

**Issue by anttikantee** [rumpkernel/rumprun#14]
For example on void linux with libssp-devel installed:

```
#    create  libc/fgets_chk.d
CC=/root/rumprun/platform/baremetal/rumptools/bin/i486--netbsdelf-gcc /root/rumprun/platform/baremetal/rumptools/bin/nbmkdep -f fgets_chk.d.tmp  --   -std=gnu99    -D__NetBSD__ -U_FORTIFY_SOURCE -Ulinux -U__linux -U__linux__ -U__gnu_linux__ --sysroot=/root/rumprun/platform/baremetal/rumptools/dest  -D_LIBC -DLIBC_SCCS -DSYSLIBC_SCCS -D_REENTRANT -D_DIAGNOSTIC -DMLIBDIR=\"\" -DHESIOD -DINET6 -DNLS -DYP -I/root/rumprun/src-netbsd/lib/libc/include -I/root/rumprun/src-netbsd/lib/libc -I/root/rumprun/src-netbsd/sys -I/root/rumprun/src-netbsd/lib/libc/compat/../locale -I/root/rumprun/src-netbsd/lib/libc/compat/stdlib -I/root/rumprun/src-netbsd/lib/libc/compat/../stdlib -D__BUILD_LEGACY -I/root/rumprun/src-netbsd/lib/libc/../../common/lib/libc/quad -I/root/rumprun/src-netbsd/lib/libc/../../common/lib/libc/string -I/root/rumprun/src-netbsd/lib/libc/../../common/lib/libc/arch/i386/string -D__DBINTERFACE_PRIVATE -I/root/rumprun/src-netbsd/libexec/ld.elf_so -I/root/rumprun/src-netbsd/lib/libc/dlfcn -I/root/rumprun/src-netbsd/lib/libc/gdtoa -I/root/rumprun/src-netbsd/lib/libc/locale -DNO_FENV_H -I/root/rumprun/src-netbsd/lib/libc/arch/i386/gdtoa -DWITH_RUNE -I/root/rumprun/src-netbsd/lib/libc -DPOSIX_MISTAKE -DCOMPAT__RES -DUSE_POLL -DPORTMAP -DWIDE_DOUBLE -DALL_STATE -DUSG_COMPAT      /root/rumprun/src-netbsd/lib/libc/ssp/fgets_chk.c &&  mv fgets_chk.d.tmp fgets_chk.d
In file included from /root/rumprun/src-netbsd/lib/libc/ssp/fgets_chk.c:41:0:
/usr/lib/gcc/x86_64-unknown-linux-gnu/4.9/include/ssp/stdio.h:38:17: fatal error: ssp.h: No such file or directory
 #include <ssp.h>
                 ^
compilation terminated.
nbmkdep: compile failed.
```

For whatever reason, gcc doesn't pick up the ssp headers from --sysroot, but rather from the host.

cc @xtraeme

**Comment by fltt** [comment]
Hello. I had a similar issue with gcc on FreeBSD. gcc has a number of "fixed" (and other) headers installed under /usr/[local/]lib/[gcc&lt;version&gt;/]gcc/&lt;arch&gt;/&lt;version&gt;/ that have precedence over (supposedly broken) system headers, even when --sysroot is in use.
I've fixed this issue adding "-isystem =/usr/include" to the "BUILDRUMP_CPPFLAGS=--sysroot=\${BUILDRUMP_STAGE}" line in the "buildrump.sh" script -- it gives precedence to the specified directory over the directories containing the fixed headers.

**NOTE**: The '=' sign is not the assignment operator but a placeholder for the sysroot directory, so don't put an extra space after it. You can, however, remove the space before it, but I think that that would make the '=' sign look more like an assignment operator.

**Comment by anttikantee** [comment]
Yes, I remember having such problems on FreeBSD (and "solving" them by temporarily renaming the fixed header directory during the build test ;)

Did you also test building rumpctrl with a buildrump.sh containing your patch?  If so, please submit a pullup to buildrump.sh (adding yourself to AUTHORS, if you so choose to do).

**Comment by fltt** [comment]
Done. See rumpkernel/buildrump.sh#88

At first rebuilding buildrump.sh failed because of some undefined pthread function. I had both Clang and GCC installed and when buildrump.sh was run, it used both. Is that the expected behaviour? Maybe another patch is in order. However, without GCC around both buildrump.sh and rumpctrl's buildnb.sh complete without issues.

**Comment by fltt** [comment]
Oh! I only tested it on 32 bit FreeBSD (I don't have a Linux box at hand, at the moment).

**Comment by anttikantee** [comment]
No, the undefined pthread stuff is not expected.  Please file an issue with more details.
(I'll look at/merge your patches later today. Thanks!)

**Comment by fltt** [comment]
Actually my question was about if it was expected for the `build-rr.sh` to use both Clang and gcc in the same run.
However I have to invest more time on this issue. As soon as I've a better understanding of the problem I'll give you some feedback.

**Comment by anttikantee** [comment]
No, using both clang and gcc isn't expected -- at least not as far as I can remember -- and actually clang for Rumprun isn't fully supported.  Clang doesn't support the gcc specs files, which are used by the application wrapper toolchain.  However, you should be able to build Rumprun itself (or not, depending on how far off my WAG is landing).

**Comment by anttikantee** [comment]
oh, the app-tools dir is the code reference to the above, in case that wasn't clear

**Comment by fltt** [comment]
Ops, another misunderstanding. When I tested `buildrump.sh` I did it not as part of building the Rumprun kernel, but to build a Rump kernel. In other words I've invoked it directly (without options or environment variables, that is exactly as specified in the tutorial pages of the wiki) instead of indirectly, through the `build-rr.sh` script.

However, as I said, I haven't spent enough time in analyzing the issue.
